homeassistant/components/directv/media_player.py

```diff
@@ -116,62 +123,75 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+
             # Make sure that this device is not already configured
             # Comparing based on host (IP) and clientAddr.
             if (host, loc["clientAddr"]) in known_devices:
                 _LOGGER.debug(
                     "Discovered device %s on host %s with "
                     "client address %s is already "
                     "configured",
-                    str.title(loc["locationName"]),
+                    loc_name,
                     host,
                     loc["clientAddr"],
                 )
             else:
                 _LOGGER.debug(
                     "Adding discovered device %s with client address %s",
-                    str.title(loc["locationName"]),
+                    loc_name,
                     loc["clientAddr"],
                 )
-                hosts.append(
-                    [
-                        str.title(loc["locationName"]),
-                        host,
-                        DEFAULT_PORT,
-                        loc["clientAddr"],
-                    ]
+                dtv = DIRECTV(host, DEFAULT_PORT, loc["clientAddr"])
+                devices.append(
+                    DirecTvDevice(
+                        loc_name, loc["clientAddr"], dtv, discovery_dtv_version,
+                    )
                 )
+                known_devices.add((host, loc["clientAddr"]))
 
-    dtvs = []
+    add_entities(devices)
 
-    for host in hosts:
-        dtvs.append(DirecTvDevice(*host))
-        hass.data.setdefault(DATA_DIRECTV, set()).add((host[1], host[3]))
 
-    add_entities(dtvs)
+def _get_receiver_version(client):
+    """Return the version of the DirectTV receiver."""
+    try:
+        return client.get_version()
+    except requests.exceptions.RequestException as ex:
+        _LOGGER.debug("Request exception %s trying to get receiver version", ex)
+        return None
 
 
 class DirecTvDevice(MediaPlayerDevice):
     """Representation of a DirecTV receiver on the network."""
 
-    def __init__(self, name, host, port, device):
+    def __init__(self, name, device, dtv, version_info=None):
         """Initialize the device."""
-
-        self.dtv = DIRECTV(host, port, device)
+        self.dtv = dtv
+        self._client_addr = device
         self._name = name
+        self._unique_id = None
         self._is_standby = True
         self._current = None
         self._last_update = None
         self._paused = None
         self._last_position = None
         self._is_recorded = None
-        self._is_client = device != "0"
+        self._is_client = self._client_addr != "0"
         self._assumed_state = None
         self._available = False
         self._first_error_timestamp = None
+        self._receiver_id = None
+        self._version_info = version_info
+
+        if self._version_info:
+            self._receiver_id = "".join(self._version_info.get("receiverId").split())
+            self._unique_id = f"{self._receiver_id}-{self._client_addr}"
```

Comment:
Is the client address stable between restarts and and reconfiguration etc?

Reply:
yep, its effectively the MAC address of the client device with semicolons

Reply:
For the host it is 0, it says in the description. So for clients it's their Mac address but for the host it is 0? What if we change host, what happens then?

Reply:
Is the receiver_id unique and stable per host?

Reply:
@MartinHjelmare there's no real documentation as it's a more private API for installers and integration partners but the receiver id has remained stable in my usage. The reciever is is also what's used for SSDP serial number. I'm pretty sure it's a hard coded id that DirecTV assigns to each receiver it produces. I think most people would have 1 actual host unit in a household since the genie unit handles multiple client devices. The reasons for having more than 1 host unit is going to be either storage space or number of rooms I believe.

The only case I can think for not using the receiver id in client unique maybe to handle a case where the host unit is replaced but clients remain the same. I guess we could use receiver id for the host and the mac address for clients which would allow client devices to remain consistent across host unit replacements.

Reply:
Yes, I think that change sounds good. Please do that.